src/openms/include/OpenMS/KERNEL/AnnotatedMSRawData.h

```diff
@@ -0,0 +1,153 @@
+// ETH Zurich, and Freie Universitaet Berlin 2002-2021.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: David Voigt $
+// $Authors: David Voigt $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/OpenMSConfig.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <boost/range/combine.hpp>
+
+#include <vector>
+
+namespace OpenMS
+{
+  class PeptideIdentification;
+  class MSSpectrum;
+
+  class OPENMS_DLLAPI AnnotatedMSRawData : public MSExperiment
+  {
+  public:
+    typedef std::pair<MSSpectrum&, std::vector<PeptideIdentification>&> Mapping;
+    typedef std::pair<const MSSpectrum&, const std::vector<PeptideIdentification>&> ConstMapping;
+
+    /// Default constructor
+    AnnotatedMSRawData() = default;
+    /// Move constructor for efficiently loading a MSExperiment without a deep copy.
+    explicit AnnotatedMSRawData(MSExperiment&& experiment) : MSExperiment(std::move(experiment)) {};
+    AnnotatedMSRawData(AnnotatedMSRawData&&) = default;
+    ~AnnotatedMSRawData() = default;
+
+    /// Get the peptide identifications for a single spectrum.
+    std::vector<PeptideIdentification>& getPeptideIds();
+    /// Get all peptide identifications for all spectra.
+    std::vector<std::vector<PeptideIdentification>>& getAllPeptideIds();
```

Comment:
I'm also not sure how to handle indexing. I was thinking to use a 2d vector

```
std::vector<std::vector<PeptideIdentification>>
```

where, as you mentioned, the indices of spectra in the MSExperiment correspond to the indices of the outter vector. That means however that in order to add or edit peptide ids of a specific spectrum, you'd have to know the index. I can imagine that there are cases where the index is unknown. As of now peptide identification is stored within the spectra itself so indices are not used for storing these IDs, at least not directly. Also using indices feels a bit unsafe since accessing the correct spectrum cannot be guarantied by the interface itself but must be ensured by the programmer himself.

Maybe it would be better to use a `std::map<MSSpectrum, std::vector<PeptideIdentification>>` so accessing IDs is more straight forward. This would come at the disadvantage that upon the construction of `AnnotatedMSRawData` all spectra would need to be copied of into the map and the spectra vector in `MSExperiment` might no longer be the same as the keys in the map as things are added to the map. That means you'd have to prohibit accessing the `MSExperiment::getSpectra` function somehow.\
Alternatively it might be possible to use a `std::reference_wraper` as key instead. This would complicate adding or removing spectra though because the map and actual vector of spectra always would need to be kept in sync. 

What are your thoughts on this?

Reply:
my take on what could work:
provide only std::vector<PeptideIdentification> - one PeptideIdentification (potentially empty) per spectrum
I think the vector of vector is not needed for all use cases we have (though the original design supported it)